### PR TITLE
cli/upgrade: Initial support for offline upgrades

### LINKED ISSUE #45: Offline Updates
This actually feels fairly straight forward.

Reference: https://www.freedesktop.org/software/systemd/man/latest/systemd.offline-updates.html

1. Get all the packages downloaded (eopkg up -f does this)
2. Create a symlink `/system-update > /var/cache/eopkg/packages/`
3. On reboot systemd checks for this file `/system-update` and boots into a minimal state and runs a systemd file to do the upgrade

The systemd file needs to:
1. Check the symlink points to the eopkg package cache folder
2. Do the update
3. Send a systemctl reboot

Notes:
* The update service should be after (wants?) sysinit.target
* eopkg will need an option to not fetch the latest metadata (I believe this was the default previously). This is important as it's an offline update, there is no network in the offline-update system and eopkg doesn't do the upgrade if it fails to get the latest metadata.

I'm looking into what the systemd service file looks like soon.

### PR DESCRIPTION
**Pull request by joebonrichie**
https://www.freedesktop.org/software/systemd/man/latest/systemd.offline-updates.html

Resolves #45.

### CODE REVIEW

**Comment by Justinzobel** [comment]
Is this ready for testing?

**Comment by Justinzobel** [comment]
Not listed in `eopkg help` ?

File: pisi/operations/upgrade.py [hunk 1]
```diff
@@ -207,6 +214,47 @@ def upgrade(A=[], repo=None):
...
+    #
+    # For a high-level overview
+    # 1. Fetch upgrades with eopkg up --fetch-only
+    # 2. Create symlink to /system-update to instruct systemd to perform offline upgrades
+    # 3. Soft reboot to initiate offline upgrades
+    # 3. eopkg-offline-upgrades.service will run eopkg up --bypass-update-repo --yes-all
+    # 4. System reboots and /system-update file is removed
+
+    # a bit of a hard coded safety hack here
+    if not os.path.exists('/usr/lib/systemd/system/system-update.target.wants/eopkg-offline-update.service'):
```
Comment (Justinzobel) [inline pisi/operations/upgrade.py:228]: Shouldn't this be checking for the normal service file, not the symlink?

Comment (Justinzobel) [reply pisi/operations/upgrade.py:228]: Or rather, both. As both should always exist on the system and the `system-update.target.wants` is only executed when offline updates are detected by the `/system-update` symlink?

**Comment by Justinzobel** [comment]
Sweet, got the offline update to work on my test laptop, very nice! It would also be nice to just get it to either A) display text showing eopkg output or B) show some sort of progress indicator and info that updates are in progress, please wait, etc.

I think we should probably also log all output from the upgrade attempt somewhere on the disk so that if an offline upgrade fails, we can direct the user to a log file so we can analyze it.

**Comment by Justinzobel** [comment]
Oh interesting, this did the updates before rebooting... :thinking: 

I pulled the latest commits and did it on my test laptop and I ran the updates, but it didn't reboot before doing the updates, it just appears to have dropped to a TTY, done the updates and then rebooted.

**Comment by Justinzobel** [comment]
Is this being abandoned?

**Comment by EbonJaeger** [comment]
I think some of the branches have been renamed, so the PR has to target one of the new ones. I don't know if that can be done without opening a new PR.

**Comment by Justinzobel** [comment]
Ah, OK. Good to know it's not abandoned. It's an excellent feature to avoid issues when users do updates and do not reboot.

**Pull request closed without merging**